driver/prom_processing.c

```diff
@@ -0,0 +1,118 @@
+#include "error.h"
+#include "prom_processing.h"
+#include "utils.h"
+
+#include "default_prom.c"
+
+#define PROM_VERSION            1
+#define PROM_NAME_MAX_LENGTH    256
+#define PROM_FIRST_ENTRY_OFFSET 5
+#define PROM_VERSION_OFFSET     4
+#define PROM_END_ENTRY_SIZE     (sizeof(struct prom_end_entry))
+
+
+
+static int validate_prom(struct prom_context *context)
+{
+    return calc_checksum16(context->buff, context->data_len) == 0;
+}
+
+
+struct prom_entry *prom_first_entry(struct prom_context *context)
+{
+    return (struct prom_entry *) &context->buff[PROM_FIRST_ENTRY_OFFSET];
+}
+
+
+struct prom_entry *prom_next_entry(struct prom_entry *entry)
+{
+    if (entry->tag == PROM_END_TAG)
+        return entry;
+    return (struct prom_entry *) (((char *) entry) + entry->size + 2);
+}
+
+
+struct prom_entry *prom_find_entry(struct prom_context *context, int index)
+{
+    int i = 0;
+    struct prom_entry *pentry;
+    prom_for_each_entry(pentry, context) {
+        if (i == index)
+            return pentry;
+        i++;
+    }
+    return NULL;
+}
+
+
+static int check_magic(const u32 magic)
+{
+    char *buff = (char *) &magic;
+    return buff[0] == 'D' && buff[1] == 'I' && buff[2] == 'A' && buff[3] == 'G';
+}
+
+
+void release_prom_context(struct prom_context *context)
+{
+    kfree(context);
+}
+
+
+void read_prom(struct prom_context *context, char *buff)
+{
+    for (int i = 0; i < PROM_MAX_LENGTH; i+=4)
+    {
+        u32 rval = ioread32(context->base + i);
+        memcpy(buff + i, (char *) &rval, 4);
+    }
+}
+
+
+struct prom_context *load_prom(void __iomem *base)
+{
+    int rc = 0;
+    struct prom_context *context = kzalloc(sizeof(struct prom_context), GFP_KERNEL);
+    TEST_PTR(context, rc, no_memory, "Unable to allocate PROM buffer");
+
+    context->base = base;
+
+    u32 rval = ioread32(base);
+    if (!check_magic(rval))
+    {
+        printk(KERN_INFO "PROM memory not found, falling back to default");
+        memcpy(context->buff, default_prom, sizeof(default_prom));
+        context->nentries = default_prom_nentries;
+        context->data_len = default_prom_data_len;
+        return context;
+    }
+
+    read_prom(context, context->buff);
+
+    TEST_OK(context->buff[PROM_VERSION_OFFSET] == PROM_VERSION,
+        rc = -EIO, no_version, "PROM version is not supported");
+
+    int ent_i = PROM_FIRST_ENTRY_OFFSET;
+    while (ent_i < PROM_MAX_LENGTH && context->buff[ent_i] != PROM_END_TAG)
+    {
+        ent_i += context->buff[ent_i + 1] + 2;
+        context->nentries++;
+    }
+
+    // At this point we've either run off the end of memory, or we're sitting
+    // on a putative end marker
+    TEST_OK(
+        ent_i < PROM_MAX_LENGTH - PROM_END_ENTRY_SIZE
+            && context->buff[ent_i] == PROM_END_TAG,
```

Comment:
You never validate the length of the end entry!

Reply:
Done